### PR TITLE
[vscoq2] understand Proof using

### PR DESCRIPTION
**Pull request by gares**

### CODE REVIEW

**Comment by gares** [comment]
@maximedenes early observation: the fact that the schedule does not store the ast for proof blocks is a problem again, this time I did not want to change the type (to take not only its state but the entire document) so I added the ast along with the ids. I'll refine the code, but that ast is there to stay, IMO.

File: language-server/tests/d_tests.ml [hunk 1]
```diff
@@ -17,14 +25,42 @@ let %test_unit "document: validate" =
...
+let %test_unit "document: invalidate proof 2" =
+  let dm, _ = openDoc ~uri:"foo" ~text:"Definition x := 3. Lemma foo : x = 3. Proof. reflexivity. Qed." in
+  let doc = DocumentManager.Internal.document dm in
+  let invalid, doc = Document.validate_document doc in
+  [%test_eq: sentence_id list] [] (Stateid.Set.elements invalid);
+  let s1,(s2,(s3,(s4,(s5,())))) = d_sentences doc (P(P(P(P(P(O)))))) in
+  let doc = Document.apply_text_edits doc [Document.Internal.range_of_sentence_id doc s3.id,""] in
+  let invalid, doc = Document.validate_document doc in
+  let r1,(r2,(r3,(r4,()))) = d_sentences doc (P(P(P(P(O))))) in
+  (* HERE THE BUG *)
```
Comment (gares) [inline language-server/tests/d_tests.ml:65]: @maximedenes @Hjaltesorgenfrei here a test reproducing the bug #452
Apparently even after zapping s3 the parser returns 5 sentences.

Comment (Hjaltesorgenfrei) [reply language-server/tests/d_tests.ml:65]: Yes, that seems like the bug I observed.

File: language-server/dm/executionManager.ml [hunk 2]
```diff
@@ -185,11 +227,14 @@ let handle_event event state =
...
         | Some (ProofJob.UpdateExecStatus(id,v)) ->
-            match SM.find id state.of_sentence with
-            | (Delegated (_,completion), fl) ->
+            match SM.find id state.of_sentence, v with
+            | (Delegated (_,completion), fl), _ ->
                 Option.default ignore completion v;
                 Some (update_all id (Done v) fl state)
-            | (Done _, fl) -> None (* TODO: is this possible? *)
+            | (Done (Success s), fl), Error (err,_) ->
+                (* Qed updated by worker that the proof is not complete *)
```
Comment (maximedenes) [inline language-server/dm/executionManager.ml:235]: The comment is unclear to me.

File: language-server/dm/scheduler.ml [hunk 3]
```diff
@@ -126,30 +139,58 @@ Qed.
...
 (* FIXME handle commands with side effects followed by `Abort` *)
-let push_state id ast synterp_st classif st =
+
+let find_proof_using (ast : ast) =
+  match ast.CAst.v.expr with
+  | VernacPure(VernacProof(_,Some e)) -> Some e
+  | _ -> log "no ast for proof using, looking at a default";
+         Proof_using.get_default_proof_using ()
+
+(* There is also a #[using] annotation on the proof opener we should take into account *)
```
Comment (maximedenes) [inline language-server/dm/scheduler.ml:149]: Can you put an explicit TODO here?

File: language-server/dm/scheduler.ml [hunk 4]
```diff
@@ -126,30 +139,58 @@ Qed.
...
+  | VtKeep (VtKeepAxiom | VtKeepOpaque) ->
+      if section_depth = 0 then Some Vernacexpr.SsEmpty
+      else find_proof_using_annotation block
+
+let push_state id ast synterp classif st =
   let open Vernacextend in
+  let ex_sentence = { id; ast; synterp; error_recovery = RSkip } in
   match classif with
-  | VtStartProof _ -> base_id st, open_proof_block id st, Exec(id,ast,synterp_st)
-  | VtQed (VtKeep (VtKeepAxiom | VtKeepOpaque)) when st.section_depth = 0 -> (* TODO do not delegate if command with side effect inside the proof or nested lemmas *)
```
Comment (maximedenes) [inline language-server/dm/scheduler.ml:133]: Is the TODO done?

**Review by maximedenes — APPROVED** [review]

**Pull request merged**